http.go

```diff
@@ -14,110 +14,115 @@
 package main
 
 import (
+	"bytes"
+	"compress/gzip"
 	"context"
-	"fmt"
+	"errors"
 	"io"
+	"io/ioutil"
 	"net"
 	"net/http"
-	"net/http/httputil"
-	"net/url"
 	"strconv"
 	"strings"
 
-	"golang.org/x/net/context/ctxhttp"
-
-	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	dto "github.com/prometheus/client_model/go"
 	"github.com/prometheus/common/expfmt"
 	"github.com/prometheus/common/log"
 )
 
-func (c httpConfig) GatherWithContext(ctx context.Context, r *http.Request) prometheus.GathererFunc {
-	return func() ([]*dto.MetricFamily, error) {
-		qvs := r.URL.Query()
-		qvs["module"] = qvs["module"][1:]
+const (
+	// Msg to send in response body when verification of proxied server
+	// response is failed
+	VerificationErrorMsg = "Internal Server Error: " +
+		"Response from proxied server failed verification. " +
+		"See server logs for details"
+)
 
-		url, err := url.Parse(c.Path)
-		uvs := url.Query()
-		for k, vs := range uvs {
-			for _, v := range vs {
-				qvs.Add(k, v)
-			}
-		}
+type VerifyError struct {
+	msg   string
+	cause error
+}
 
-		url.Host = net.JoinHostPort(c.Address, strconv.Itoa(c.Port))
-		url.Scheme = c.Scheme
-		url.RawQuery = qvs.Encode()
+func (e *VerifyError) Error() string { return e.msg + ": " + e.cause.Error() }
+func (e *VerifyError) Unwrap() error { return e.cause }
 
-		resp, err := ctxhttp.Get(ctx, c.httpClient, url.String())
-		if err != nil {
-			log.Errorf("http proxy for module %v failed %+v", c.mcfg.name, err)
-			proxyErrorCount.WithLabelValues(c.mcfg.name).Inc()
-			if err == context.DeadlineExceeded {
-				proxyTimeoutCount.WithLabelValues(c.mcfg.name).Inc()
-			}
-			return nil, err
-		}
-		defer resp.Body.Close()
+func (cfg moduleConfig) getReverseProxyDirectorFunc() func(*http.Request) {
+	return func(r *http.Request) {
+		vs := r.URL.Query()
+		vs["module"] = vs["module"][1:]
+		r.URL.RawQuery = vs.Encode()
+
+		r.URL.Scheme = cfg.HTTP.Scheme
+		r.URL.Host = net.JoinHostPort(cfg.HTTP.Address, strconv.Itoa(cfg.HTTP.Port))
+		r.URL.Path = cfg.HTTP.Path
+	}
+}
 
+func (cfg moduleConfig) getReverseProxyModifyResponseFunc() func(*http.Response) error {
+	return func(resp *http.Response) error {
 		if resp.StatusCode != 200 {
```

Comment:
I may be missing it, but can we not log non-200 responses here? Are the logged elsewhere? Without it, that could make debugging things a bit tricky.